### PR TITLE
pdns_notify: Support hostname for notification

### LINKED ISSUE #7319: pdns_notify  core dumps on invalid ip
- Program: pdns_notify 
 - Issue type: Bug report 

### Short description
pdns_notify aborts with coredump on usage error

### Environment
 - Operating system: OpenBSD
 - Software version: 4.1.5
 - Software source: Operating system repository

### Steps to reproduce
1. pdsn_notify 1.2.3. a

### Expected behaviour
Error message

### Actual behaviour
terminating with uncaught exception of type PDNSException
Abort trap (core dumped) 
### Usecase
A simple typo made should produce an appropriate error message.

### Description

Provide proper error message and return non-zero.

### PR DESCRIPTION
**Pull request by cmouse**
### Short description
Fixes #7319

### Checklist

I have:
- [x] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [x] compiled this code
- [x] tested this code
- [ ] included documentation (including possible behaviour changes)
- [ ] documented the code
- [ ] added or modified regression test(s)
- [ ] added or modified unit test(s)

### CODE REVIEW

File: pdns/notify.cc [hunk 1]
```diff
@@ -76,39 +77,63 @@ try
...
+     if (getaddrinfo(parts[0].c_str(), parts[1].c_str(), NULL, &info) < 0)
+       throw runtime_error("Cannot resolve '" + string(argv[1]) +"'");
+     for(auto ptr = info; ptr != NULL; ptr = ptr->ai_next)
+       addrs.emplace(ComboAddress{ptr->ai_addr, ptr->ai_addrlen});
+  }
 
+  for(const auto &addr: addrs) {
+    if (sock > -1)
+      (void)close(sock);
+    sock = socket(addr.sin4.sin_family, SOCK_DGRAM, 17);
```
Comment (omoerbeek) [inline pdns/notify.cc]: Magic number 17 for UDP, I think 0 would do

File: pdns/notify.cc [hunk 2]
```diff
@@ -139,20 +138,23 @@ try
...
+      cerr<<"Unable to receive notification response from "<<addr.toStringWithPort()<<": EOF"<<endl;
       continue;
     }
 
     string packet(buffer, len);
     MOADNSParser mdp(false, packet);
 
-    cerr<<addr.toString()<<": Received notification response with error: "<<RCode::to_s(mdp.d_header.rcode)<<endl;
+    if (mdp.d_header.rcode == 0)
+      cerr<<"Successfully notified "<<addr.toStringWithPort()<<endl;
```
Comment (omoerbeek) [inline pdns/notify.cc]: I think some failures are not processed correctly. Compare 
pdns_notify ns1.drijf.net sillydomainname
and
pdns_notify ns1.drijf.net sillydomainname,
and 
pdns_notify ns1.xs4all.nl sillydomainname

Don't know yet why ns1.drijf,net response is interpreted as succes in some cases.

Comment (cmouse) [reply pdns/notify.cc]: did you mean `ns1.drijf,net` or `ns1.drijf.net` on the last sentence?

Comment (cmouse) [reply pdns/notify.cc]: Um. I cannot understand the problem here. I tested with wireshark listening on the results and it seems your nameserver is sending NO ERROR result code for the first notify followed with SERVERROR bit later. 

```
Frame 5: 75 bytes on wire (600 bits), 75 bytes captured (600 bits) on interface 0
Ethernet II, Src: 52:54:00:c9:28:a8, Dst: 00:00:5e:00:01:23
Internet Protocol Version 4, Src: 95.175.99.157, Dst: 94.142.244.34
User Datagram Protocol, Src Port: 57416, Dst Port: 53
Domain Name System (query)
    Transaction ID: 0x6745
    Flags: 0x2000 Zone change notification
        0... .... .... .... = Response: Message is a query
        .010 0... .... .... = Opcode: Zone change notification (4)
        .... ..0. .... .... = Truncated: Message is not truncated
        .... ...0 .... .... = Recursion desired: Don't do query recursively
        .... .... .0.. .... = Z: reserved (0)
        .... .... ...0 .... = Non-authenticated data: Unacceptable
    Questions: 1
    Answer RRs: 0
    Authority RRs: 0
    Additional RRs: 0
    Queries
        sillydomainname: type SOA, class IN
            Name: sillydomainname
            [Name Length: 15]
            [Label Count: 1]
            Type: SOA (Start Of a zone of Authority) (6)
            Class: IN (0x0001)

Frame 6: 75 bytes on wire (600 bits), 75 bytes captured (600 bits) on interface 0
Ethernet II, Src: 80:ac:ac:d7:87:f0, Dst: 52:54:00:c9:28:a8
Internet Protocol Version 4, Src: 94.142.244.34, Dst: 95.175.99.157
User Datagram Protocol, Src Port: 53, Dst Port: 57416
Domain Name System (response)
    Transaction ID: 0x6745
    Flags: 0xa400 Zone change notification response, No error
        1... .... .... .... = Response: Message is a response
        .010 0... .... .... = Opcode: Zone change notification (4)
        .... .1.. .... .... = Authoritative: Server is an authority for domain
        .... ..0. .... .... = Truncated: Message is not truncated
        .... ...0 .... .... = Recursion desired: Don't do query recursively
        .... .... 0... .... = Recursion available: Server can't do recursive queries
        .... .... .0.. .... = Z: reserved (0)
        .... .... ..0. .... = Answer authenticated: Answer/authority portion was not authenticated by the server
        .... .... ...0 .... = Non-authenticated data: Unacceptable
        .... .... .... 0000 = Reply code: No error (0)
    Questions: 1
    Answer RRs: 0
    Authority RRs: 0
    Additional RRs: 0
    Queries
        sillydomainname: type SOA, class IN
            Name: sillydomainname
            [Name Length: 15]
            [Label Count: 1]
            Type: SOA (Start Of a zone of Authority) (6)
            Class: IN (0x0001)
    [Request In: 5]
    [Time: 0.040702528 seconds]

Frame 7: 95 bytes on wire (760 bits), 95 bytes captured (760 bits) on interface 0
Ethernet II, Src: 52:54:00:c9:28:a8, Dst: 80:ac:ac:d7:87:f0
Internet Protocol Version 6, Src: 2a03:e581:4:200::10, Dst: 2a02:898::216:1
User Datagram Protocol, Src Port: 59999, Dst Port: 53
Domain Name System (query)
    Transaction ID: 0xc623
    Flags: 0x2000 Zone change notification
        0... .... .... .... = Response: Message is a query
        .010 0... .... .... = Opcode: Zone change notification (4)
        .... ..0. .... .... = Truncated: Message is not truncated
        .... ...0 .... .... = Recursion desired: Don't do query recursively
        .... .... .0.. .... = Z: reserved (0)
        .... .... ...0 .... = Non-authenticated data: Unacceptable
    Questions: 1
    Answer RRs: 0
    Authority RRs: 0
    Additional RRs: 0
    Queries
        sillydomainname: type SOA, class IN
            Name: sillydomainname
            [Name Length: 15]
            [Label Count: 1]
            Type: SOA (Start Of a zone of Authority) (6)
            Class: IN (0x0001)

Frame 8: 95 bytes on wire (760 bits), 95 bytes captured (760 bits) on interface 0
Ethernet II, Src: 80:ac:ac:d7:87:f0, Dst: 52:54:00:c9:28:a8
Internet Protocol Version 6, Src: 2a02:898::216:1, Dst: 2a03:e581:4:200::10
User Datagram Protocol, Src Port: 53, Dst Port: 59999
Domain Name System (response)
    Transaction ID: 0xc623
    Flags: 0xa400 Zone change notification response, No error
        1... .... .... .... = Response: Message is a response
        .010 0... .... .... = Opcode: Zone change notification (4)
        .... .1.. .... .... = Authoritative: Server is an authority for domain
        .... ..0. .... .... = Truncated: Message is not truncated
        .... ...0 .... .... = Recursion desired: Don't do query recursively
        .... .... 0... .... = Recursion available: Server can't do recursive queries
        .... .... .0.. .... = Z: reserved (0)
        .... .... ..0. .... = Answer authenticated: Answer/authority portion was not authenticated by the server
        .... .... ...0 .... = Non-authenticated data: Unacceptable
        .... .... .... 0000 = Reply code: No error (0)
    Questions: 1
    Answer RRs: 0
    Authority RRs: 0
    Additional RRs: 0
    Queries
        sillydomainname: type SOA, class IN
            Name: sillydomainname
            [Name Length: 15]
            [Label Count: 1]
            Type: SOA (Start Of a zone of Authority) (6)
            Class: IN (0x0001)
    [Request In: 7]
    [Time: 0.042313931 seconds]

Frame 9: 75 bytes on wire (600 bits), 75 bytes captured (600 bits) on interface 0
Ethernet II, Src: 80:ac:ac:d7:87:f0, Dst: 52:54:00:c9:28:a8
Internet Protocol Version 4, Src: 94.142.244.34, Dst: 95.175.99.157
User Datagram Protocol, Src Port: 18244, Dst Port: 53
Domain Name System (query)
    Transaction ID: 0x5432
    Flags: 0x0000 Standard query
        0... .... .... .... = Response: Message is a query
        .000 0... .... .... = Opcode: Standard query (0)
        .... ..0. .... .... = Truncated: Message is not truncated
        .... ...0 .... .... = Recursion desired: Don't do query recursively
        .... .... .0.. .... = Z: reserved (0)
        .... .... ...0 .... = Non-authenticated data: Unacceptable
    Questions: 1
    Answer RRs: 0
    Authority RRs: 0
    Additional RRs: 0
    Queries
        sillydomainname: type SOA, class IN
            Name: sillydomainname
            [Name Length: 15]
            [Label Count: 1]
            Type: SOA (Start Of a zone of Authority) (6)
            Class: IN (0x0001)

Frame 10: 95 bytes on wire (760 bits), 95 bytes captured (760 bits) on interface 0
Ethernet II, Src: 80:ac:ac:d7:87:f0, Dst: 52:54:00:c9:28:a8
Internet Protocol Version 6, Src: 2a02:898::216:1, Dst: 2a03:e581:4:200::10
User Datagram Protocol, Src Port: 10460, Dst Port: 53
Domain Name System (query)
    Transaction ID: 0x2f25
    Flags: 0x0000 Standard query
        0... .... .... .... = Response: Message is a query
        .000 0... .... .... = Opcode: Standard query (0)
        .... ..0. .... .... = Truncated: Message is not truncated
        .... ...0 .... .... = Recursion desired: Don't do query recursively
        .... .... .0.. .... = Z: reserved (0)
        .... .... ...0 .... = Non-authenticated data: Unacceptable
    Questions: 1
    Answer RRs: 0
    Authority RRs: 0
    Additional RRs: 0
    Queries
        sillydomainname: type SOA, class IN
            Name: sillydomainname
            [Name Length: 15]
            [Label Count: 1]
            Type: SOA (Start Of a zone of Authority) (6)
            Class: IN (0x0001)

Frame 11: 95 bytes on wire (760 bits), 95 bytes captured (760 bits) on interface 0
Ethernet II, Src: 52:54:00:c9:28:a8, Dst: 80:ac:ac:d7:87:f0
Internet Protocol Version 6, Src: 2a03:e581:4:200::10, Dst: 2a02:898::216:1
User Datagram Protocol, Src Port: 53, Dst Port: 10460
Domain Name System (response)
    Transaction ID: 0x2f25
    Flags: 0x8005 Standard query response, Refused
        1... .... .... .... = Response: Message is a response
        .000 0... .... .... = Opcode: Standard query (0)
        .... .0.. .... .... = Authoritative: Server is not an authority for domain
        .... ..0. .... .... = Truncated: Message is not truncated
        .... ...0 .... .... = Recursion desired: Don't do query recursively
        .... .... 0... .... = Recursion available: Server can't do recursive queries
        .... .... .0.. .... = Z: reserved (0)
        .... .... ..0. .... = Answer authenticated: Answer/authority portion was not authenticated by the server
        .... .... ...0 .... = Non-authenticated data: Unacceptable
        .... .... .... 0101 = Reply code: Refused (5)
    Questions: 1
    Answer RRs: 0
    Authority RRs: 0
    Additional RRs: 0
    Queries
        sillydomainname: type SOA, class IN
            Name: sillydomainname
            [Name Length: 15]
            [Label Count: 1]
            Type: SOA (Start Of a zone of Authority) (6)
            Class: IN (0x0001)
    [Request In: 10]
    [Time: 0.000180666 seconds]

Frame 12: 75 bytes on wire (600 bits), 75 bytes captured (600 bits) on interface 0
Ethernet II, Src: 80:ac:ac:d7:87:f0, Dst: 52:54:00:c9:28:a8
Internet Protocol Version 4, Src: 94.142.244.34, Dst: 95.175.99.157
User Datagram Protocol, Src Port: 53, Dst Port: 57416
Domain Name System (response)
    Transaction ID: 0x6745
        [Expert Info (Warning/Protocol): DNS response retransmission. Original response in frame 6]
            [DNS response retransmission. Original response in frame 6]
            [Severity level: Warning]
            [Group: Protocol]
    Flags: 0xa402 Zone change notification response, Server failure
        1... .... .... .... = Response: Message is a response
        .010 0... .... .... = Opcode: Zone change notification (4)
        .... .1.. .... .... = Authoritative: Server is an authority for domain
        .... ..0. .... .... = Truncated: Message is not truncated
        .... ...0 .... .... = Recursion desired: Don't do query recursively
        .... .... 0... .... = Recursion available: Server can't do recursive queries
        .... .... .0.. .... = Z: reserved (0)
        .... .... ..0. .... = Answer authenticated: Answer/authority portion was not authenticated by the server
        .... .... ...0 .... = Non-authenticated data: Unacceptable
        .... .... .... 0010 = Reply code: Server failure (2)
    Questions: 1
    Answer RRs: 0
    Authority RRs: 0
    Additional RRs: 0
    Queries
        sillydomainname: type SOA, class IN
            Name: sillydomainname
            [Name Length: 15]
            [Label Count: 1]
            Type: SOA (Start Of a zone of Authority) (6)
            Class: IN (0x0001)
    [Retransmitted response. Original response in: 6]
    [Retransmission: True]

Frame 13: 95 bytes on wire (760 bits), 95 bytes captured (760 bits) on interface 0
Ethernet II, Src: 80:ac:ac:d7:87:f0, Dst: 52:54:00:c9:28:a8
Internet Protocol Version 6, Src: 2a02:898::216:1, Dst: 2a03:e581:4:200::10
User Datagram Protocol, Src Port: 53, Dst Port: 59999
Domain Name System (response)
    Transaction ID: 0xc623
        [Expert Info (Warning/Protocol): DNS response retransmission. Original response in frame 8]
            [DNS response retransmission. Original response in frame 8]
            [Severity level: Warning]
            [Group: Protocol]
    Flags: 0xa402 Zone change notification response, Server failure
        1... .... .... .... = Response: Message is a response
        .010 0... .... .... = Opcode: Zone change notification (4)
        .... .1.. .... .... = Authoritative: Server is an authority for domain
        .... ..0. .... .... = Truncated: Message is not truncated
        .... ...0 .... .... = Recursion desired: Don't do query recursively
        .... .... 0... .... = Recursion available: Server can't do recursive queries
        .... .... .0.. .... = Z: reserved (0)
        .... .... ..0. .... = Answer authenticated: Answer/authority portion was not authenticated by the server
        .... .... ...0 .... = Non-authenticated data: Unacceptable
        .... .... .... 0010 = Reply code: Server failure (2)
    Questions: 1
    Answer RRs: 0
    Authority RRs: 0
    Additional RRs: 0
    Queries
        sillydomainname: type SOA, class IN
            Name: sillydomainname
            [Name Length: 15]
            [Label Count: 1]
            Type: SOA (Start Of a zone of Authority) (6)
            Class: IN (0x0001)
    [Retransmitted response. Original response in: 8]
    [Retransmission: True]
```

Comment (omoerbeek) [reply pdns/notify.cc]: > did you mean `ns1.drijf,net` or `ns1.drijf.net` on the last sentence?

i meant dot

Comment (omoerbeek) [reply pdns/notify.cc]: ns1.drijf.net runs powerdns. I'll file a separate issue for this, since it is not related to the hostname change itself.

**Pull request merged**